doc/notes.1
.TH NOTES 1 "August 2004" "Newts" "Newts Reference Manual"

.SH NAME
notes \- legacy client for the Newts notesfile system

.SH SYNOPSIS
.B notes
[\fIoptions\fR] [\fINOTESFILE\fR]...

.SH DESCRIPTION
.B notes
is a client for \fBNewts\fR, a notesfile, or community message board, system.
.B notes
is designed to look like and function like the program of the same name
distributed in the Notesfiles package from the University of Illinois at
Urbana-Champaign.

The
.B notes
program can be used to read and write notes, update options for those
notesfiles in which you are a director, save copies of notes as text files, and
various other functions.

If no notesfiles are specified, the
.B NFSEQ
environment variable will be checked to determine which notesfiles to view.

.SH OPTIONS

.TP
\fB\-a\fR, \fB\-\^\-alternate\fR=\fINAME\fR
Use a named subsequencer.  Each subsequencer has its own independent set of
timestamps, and each can be used independently to sequence through the same
notesfiles.  If no other sequencer mode is specified, \fB\-\^\-sequencer\fR
will be assumed.

This option was originally used for shared accounts, so multiple users could
track their sequencers individually.  Since most sites have moved away from
using shared accounts, this option is rarely used.

.TP
\fB\-b\fR, \fB\-\^\-black\-threads\fR
When sequencing, for threads with basenotes that are blacklisted, skip all
responses in the thread as well as the basenote.  If the
\fB\-\^\-white\-basenotes\fR option is specified, this option does nothing.

.TP
\fB\-e\fR, \fB\-\^\-seq\-own\-notes\fR
With this option set, the sequencer will "trip" on notes you write.  If your
note is the only new note in a notesfile, that notesfile will be entered and
your note displayed.

.TP
\fB\-f\fR, \fB\-\^\-file\fR=\fIFILE\fR
Read a list of notesfiles to view from the specified file, instead of expecting
such a list to be provided on the command line.  This option does not imply use
of the sequencer, but can be used with the sequencer.

.TP
\fB\-g\fR, \fB\-\^\-no\-signature\fR
Do not include a signature in any notes or responses you compose, even if
a ~/.signature or similar file is located.

.TP
\fB\-h\fR, \fB\-\^\-help\fR
Print a summary of usage and command-line options for
.B notes
and exit.

.TP
\fB\-i\fR, \fB\-\^\-index\fR
Use the index sequencer.  This mode will skip notesfiles with no new notes, and
will open to the index screen of each notesfile with new notes, instead of
immediately opening to the first modified note.

.TP
\fB\-k\fR, \fB\-\^\-skip\-own\-notes\fR
With this option set, the sequencer will ignore notes you write for the purpose
of determining whether a notesfile has been updated.  This is the default
behavior for the sequencer.

.TP
\fB\-m\fR, \fB\-\^\-modern\fR
Use the "modern" display interface, featuring a more consistent user interface,
error messages, and prompts.  This is the default display mode.

.TP
\fB\-n\fR, \fB\-\^\-no\-sequencer\fR
Do not use the sequencer.  This is the default mode.

.TP
\fB\-o\fR, \fB\-\^\-time\fR=\fITIME\fR
Specify a particular date/time for the sequencer to use.  In this mode, all
notes written after the given time will be considered to be "new", and the
sequencer time will not be updated after you quit.  If no other sequencer mode
is specified, \fB\-\^\-sequencer\fR will be assumed.  Most common formats for
specifying dates and times are accepted by this option.

.TP
\fB\-s\fR, \fB\-\^\-sequencer\fR
Use the sequencer.  The standard sequencer mode will skip notesfiles with no
new notes, and will open immediately to the first basenote in each notesfile
with modified or new material.

.TP
\fB\-t\fR, \fB\-\^\-traditional\fR, \fB\-\^\-imsa\fR
Use the "traditional" display mode; this is the appearance that the Notesfile
system from the University of Illinois at Urbana-Champaign used, replicated as
closely as possible while still fixing obvious display bugs.  This option
implies \fB\-\^\-seq\-own\-notes\fR.

.TP
\fB\-u\fR, \fB\-\^\-user\fR=\fIUSERNAME\fR
As root, run the client as the specified user.  This means you'll use the
user's sequencer file, post as the user, etc; for all intents and purposes,
this option is identical to using \fBsu\fR(1) to become the user, then running
\fBnotes\fR(1), but is much easier to use when debugging.

.TP
\fB\-w\fR, \fB\-\^\-white\-basenotes\fR
Cause basenotes to be exempt from blacklisting.  With this option specified,
all basenotes will be displayed.  If both \fB\-\^\-white\-basenotes\fR and
\fB\-\^\-black\-threads\fR are specified, only this option will apply.

.TP
\fB\-x\fR, \fB\-\^\-extended\fR
Use the extended sequencer.  This mode will enter each notesfile specified,
regardless of whether there are new notes.  If a given notesfile has no new
notes, it will open to the index screen.  If a notesfile does have new notes,
the notesfile will open to the first basenote with modified or new material.

.TP
\fB\-z\fR, \fB\-\^\-no\-blacklist\fR
Do not use the blacklist; display all notes.

.TP
\fB\-\^\-debug\fR
Print debugging messages to standard error as the client runs; you will
probably want to save these messages to a file using redirection of standard
error.

.TP
\fB\-\^\-version\fR
Print version information for
.B notes
and exit.

.SH ENVIRONMENT VARIABLES
.B notes
is affected by the following environment variables.

.TP
\fBNFEDITOR\fR, \fBNFED\fR, \fBVISUAL\fR, \fBEDITOR\fR
These variables specify the user's preferred text editor; examples include
\fBemacs\fR(1), \fBvim\fR(1), and \fBnano\fR(1).

.TP
\fBNFMAILER\fR, \fBMAILER\fR
These variables specify the user's preferred mail user agent; examples include
\fBmutt\fR(1) and \fBpine(\fR1).  If you're a die-hard Emacs user, try setting
.B NFMAILER
to \fBemacs\fR.

.TP
\fBNFPAGER\fR, \fBPAGER\fR
These variables specify the user's preferred text pager; examples include
\fBless\fR(1) and \fBmore(\fR1).

.TP
\fBNFSEQ\fR
If no notesfiles are specified on the command line and this variable is set, it
will be used as the list of notesfiles to read.

.TP
\fBNFTALK\fR, \fBTALK\fR
These variables specify the user's preferred console-based user-to-user
communications program; examples include \fBytalk\fR(1) and \fBwrite\fR(1).

.TP
.B SHELL
This environment variable controls which shell
.B notes
will execute when asked to spawn a subshell.

.TP
.B TMPDIR
If this environment variable is set,
.B notes
will create all its temporary files in the specified directory.

.SH AUTHOR
.B Newts
was written and is maintained by Tyler Berry <tyler+newts@thoughtlocker.net>.

.SH SEE ALSO
\fBautoseq\fR(1), \fBchecknotes\fR(1), \fBgetnote\fR(1), \fBmknf\fR(1),
\fBnfadmin\fR(1), \fBnfdump\fR(1), \fBnfload\fR(1), \fBnfmail\fR(1),
\fBnfpipe\fR(1), \fBnfprint\fR(1), \fBnfstats\fR(1), \fBnftimestamp\fR(1),
\fBrmnf\fR(1)

The full documentation for
.B Newts
is maintained as a Texinfo manual.  If the
.B info
and
.B Newts
programs are properly installed at your site, the command
.IP
.B info newts
.PP
should give you access to the complete manual.


NEWS
Newts NEWS                                                         -*- Text -*-
-------------------------------------------------------------------------------

* ______ - Newts 0.15

- More unit tests.
- More changes and additions to the client API.
- More documentation of the client API.

* Wednesday, December 14, 2005 - Newts 0.14.8

- Include the SVN revision number in the output of --version.
- Added unit tests for the client API.
- Began adding documentation for the client API in a Doxygen-compatible way.
- Substantial changes and additions to the client API.
- Substantial refactoring in the main client, both to comply with the revised
  client API and to clean the code up in general.
- Modified the utility frontend programs to use the revised client API.
- Refactored the pager to be modular, for purposes of future rewrites to add
  things like Unicode support.
- Use gnulib-tool to generate Gnulib support.
- Automatically use the most recent version of Libtool.

* Tuesday, July 26, 2005 - Newts 0.14.7

- Fixed signal handling so that suspending a spawned process works correctly.
- Added David Mellis's python bindings to the repository and distribution.

* Friday, June 16, 2005 - Newts 0.14.6

- Fixed a bug in nfdump where the last note in a dumped notesfile was omitted.
- Added 'g' to perform one-touch seqing.

* Monday, March 21, 2005 - Newts 0.14.5

- Updated to Libtool 1.5.14.
- Fixed a segfault when trying to use the alternate sequencer.
- Made the alternate sequencer activate the sequencer automatically.
- Made the alternate sequencer work correctly with the '--user' option.
- Fixed the broken --time option, which was not actually setting the time at
  all. Oops.
- Fixed a few odd display bugs.
- Display messages are now 'Unicode-friendly'.
  See <http://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html> for details.
- Resynced with Gnulib.

* Sunday, January 23, 2005 - Newts 0.14.4

- Updated to Libtool 1.5.10.
- Corrected some mistakes in the Texinfo manual.
- Fixed a bug which would cause a segfault when replying to a note from a local
  user who had been deleted from the system, or forwarding a mail from a note
  written by such a user.
- Fixed a bug in nfload where the dumpfile would fail to open if nfload was run
  as root. (It was dropping root permissions and not picking them back up.)
- Resynced with Gnulib, which involved some fairly major updates to xmalloc and
  friends, as the interface Gnulib uses changed.

* Tuesday, September 14, 2004 - Newts 0.14.3

- Corrected a typo in nfadmin.1.
- Synchronized with new files in Gnulib.

* Tuesday, August 10, 2004 - Newts 0.14.2

- Corrected a Whole Bunch of inconsistencies between the Texinfo manual, the
  newly-developed manpages, and the actual operation of the programs.
- Added manpages for all currently-built programs.
- Renamed 'printnote' to 'getnote'.
- Removed the '-d' short option from all programs, which used to correspond to
  '--debug'.  It was blocking the use of '-d' for more important options which
  are far more likely to see use.

* Tuesday, June 29, 2004 - Newts 0.14.1

- Completely redid all compatibility code using a recent CVS version of
  Gnulib.

* Thursday, June 10, 2004 - Newts 0.14

- Adding the missing code to the nfload program which will load access entries
  from the dumpfile into the notesfile.
- Added debugging support to all of the various binaries built by the package,
  via the --debug command-line option.
- Added an option for debugging that allows the notes client to setuid to an
  arbitrary user when being run by root. Useful for running notes inside gdb.
- Changed the license for the Newts Reference Manual - it's no longer GFDL.
- Fixed a bug which would cause a segfault when reading a note from a local
  user who had been deleted from the system.
- Reorganized the entire directory structure of the source tree.
- Reorganized header files to install the relevant interfaces.
- Began systematic documentation of the libnewtsclient interface.

* Tuesday, March 9, 2004 - Newts 0.13.2

- Fixed a display bug which caused informational messages and prompts to fail
  to display when a person tried to enter a notesfile without rights.
- Syncronized with new files in Gnulib.

* Friday, January 16, 2003 - Newts 0.13.1

- Synchronized with new releases of GNU Autotools and Gnulib.

* Tuesday, December 9, 2003 - Newts 0.13

- Added nfdump, completing the set of basic utilities.
- Added some sanity checks to the 't' talk command.
- Various small bits of code cleanup.
- Made sure that root doesn't accidentally lose the ability to write files.

* Monday, September 29, 2003 - Newts 0.12.7

- Portability work for sparc-sun-solaris2.6.
- utmp scanning to determine whether an author is online.
- Synchronization and updates for portability code.

* Thursday, September 18, 2003 - Newts 0.12.6

- Added a couple missing m4 files.

* Friday, September 5, 2003 - Newts 0.12.5

- Code cleanup, fixes, and error-checking.

* Saturday, August 30, 2003 - Newts 0.12.4

- Added man pages for notes(1) and autoseq(1).
- Fixed inconsistencies between documentation and actual behavior.
- Made a couple command line options make more sense.
- Fixed distribution of a couple missing files.

* Tuesday, August 26, 2003 - Newts 0.12.3

- Fixed a mistake with an earlier fix to the 'N' key.
- Prepared for a CVS branch to add Readline support.

* Tuesday, August 19, 2003 - Newts 0.12.2

- Interim release for code maintenance.
- Lots and lots of code cleanup.
- Fixed lots of FIXMEs in the code; these were largely for incorrect usage of
  one kind or another; static buffers instead of dynamic alloc, not checking
  return values, etc. It's not there yet, but it's getting there.

* Wednesday, August 13, 2003 - Newts 0.12.1

- Staving off disaster, a bug in notesfile parsing was discovered and slain in
  the embryonic version 0.12, which is supplanted by this release.
- Finally fixed that nfload bug that's been driving me nuts for the past three

clients/notes/notes.c
/*
 * notes.c - main routine for legacy notesfile interface
 *
 * This file is part of the Newts notesfile system.
 * Copyright (C) 2003, 2004, 2005, 2006 Tyler Berry.
 *
 * Newts is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * Newts is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Newts; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include "notes.h"
#include "signals.h"

#include "curses_wrapper.h"
#include "dirname.h"
#include "getdate.h"
#include "getopt.h"
#include "strtok_r.h"
#include "vasprintf.h"

#if HAVE_LANGINFO_H
# include <langinfo.h>
#endif

#if HAVE_LOCALE_H
# include <locale.h>
#endif

#if HAVE_NCURSES_H
# include <ncurses.h>
#elif HAVE_CURSES_H
# include <curses.h>
#endif

#if HAVE_PWD_H
# include <pwd.h>
#endif

/* FIXME: should be removed. */

#if HAVE_SYS_STAT_H
# include <sys/stat.h>
#endif

#if HAVE_SYS_WAIT_H
# include <sys/wait.h>
#endif

/* Alternate sequencer name flag */
int alt_sequencer = FALSE;

/* Alternate sequencer name */
char *alt_seqname;

/* Alternate sequencer time flag */
int alt_time = FALSE;

/* Controls whether to skip entire threads with blacklisted basenotes */
int black_skip_seq = FALSE;

/* Controls whether to print out debug info to stderr. */
int debug = FALSE;

/* The name the program was called */
char *program_name;

/* Controls whether to use the blacklist */
int no_blacklist = FALSE;

/* Controls whether or not to append signatures to new notes/resps */
int signature = TRUE;

/* Original sequencer time */
time_t orig_seqtime;

/* Controls whether user's posts trip the sequencer */
int seq_own_notes = FALSE;

/* Sequencer time */
time_t seqtime;

/* If and what kind of sequencer we're using */
int sequencer = NONE;

/* Controls whether we want an old-school look */
int traditional = FALSE;

/* Controls whether to ignore blacklisting for basenotes */
int white_basenotes = FALSE;

int
main (int argc, char **argv)
{
  List nflist;

  short file_flag = FALSE;
  short quit_flag = FALSE;
  int result;

  newts_failed_malloc_hook = curses_malloc_die;

  srand ((unsigned) time (NULL));

  time (&orig_seqtime);
  time (&seqtime);

#ifdef __GLIBC__
  program_name = program_invocation_short_name;
#else
  program_name = base_name (argv[0]);
#endif

  /* Initialize i18n. */

#ifdef HAVE_SETLOCALE
  if (setlocale (LC_ALL, "") == NULL)
    {
      fprintf (stderr, _("%s: could not determine your locale\nCheck the "
                         "environment variables LANG, LC_ALL, etc.\n"),
               program_name);

      exit (EXIT_FAILURE);
    }
#endif

#if ENABLE_NLS
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif

  /* Initial setup and global variable init. */

  init_blacklist ();
  list_init (&nflist,
             (void * (*) (void)) nfref_alloc,
             (void (*) (void *)) nfref_free,
             NULL);

  setup ();

  while (1)
    {
      int opt;
      short override_flag = FALSE;

      static struct option long_options[] =
        {
          {"alternate",       required_argument, 0, 'a'},
          {"black-threads",   no_argument,       0, 'b'},
          {"debug",           no_argument,       0, 'D'},
          {"seq-own-notes",   no_argument,       0, 'e'},
          {"file",            required_argument, 0, 'f'},
          {"no-signature",    no_argument,       0, 'g'},
          {"index",           no_argument,       0, 'i'},
          {"skip-own-notes",  no_argument,       0, 'k'},
          {"modern",          no_argument,       0, 'm'},
          {"no-sequencer",    no_argument,       0, 'n'},
          {"time",            required_argument, 0, 'o'},
          {"sequencer",       no_argument,       0, 's'},
          {"traditional",     no_argument,       0, 't'},
          {"user",            required_argument, 0, 'u'},
          {"white-basenotes", no_argument,       0, 'w'},
          {"extended",        no_argument,       0, 'x'},
          {"imsa",            no_argument,       0, 'y'},
          {"no-blacklist",    no_argument,       0, 'z'},
          {"help",            no_argument,       0, 'h'},
          {"version",         no_argument,       0, 0},
          {0, 0, 0, 0}
        };

      opt = getopt_long (argc, argv, "a:bDef:ghikmno:stu:wxz",
                         long_options, NULL);

      if (opt == -1)
        break;

      switch (opt)
        {
        case 0:
          {
            char *version_string;

            asprintf (&version_string, _("revision: %s"), newts_revision);
            printf (N_("notes - %s %s (%s)\n"), PACKAGE_NAME, VERSION,
                    version_string);

            free (version_string);

            list_destroy (&nflist);
            teardown ();

            if (fclose (stdout) == EOF)
              error (EXIT_FAILURE, errno, _("error writing output"));

            exit (EXIT_SUCCESS);
          }

        case 'a':
          {
            alt_sequencer = TRUE;

            if (sequencer == NONE)
              sequencer = SEQUENCER;

            newts_nrealloc (seqname, strlen (username) + strlen (optarg) + 2,
                            sizeof (char));
            strcpy (seqname, username);
            strcat (seqname, ":");
            strcat (seqname, optarg);

            if (alt_seqname)
              newts_free (alt_seqname);
            alt_seqname = newts_strdup (optarg);

            break;
          }

        case 'b':
          black_skip_seq = TRUE;
          break;

        case 'D':
          debug = TRUE;
          break;

        case 'f':
          if (parse_file (optarg, &nflist))
            file_flag = TRUE;
          break;

        case 'g':
          signature = FALSE;
          break;

        case 'i':
          sequencer = INDEX;
          break;

        case 'k':
          seq_own_notes = FALSE;
          override_flag = TRUE;
          break;

        case 'm':
          seq_own_notes = TRUE;
          break;

        case 'n':
          sequencer = NONE;
          break;

        case 'o':
          {
            struct timespec parsed_time, now;
            now.tv_sec = orig_seqtime;

            if (get_date (&parsed_time, optarg, &now))
              {
                if (alt_time)
                  {
                    fprintf (stderr, _("%s: cannot specify multiple alternate times\n"),
                             program_name);
                    fprintf (stderr, _("See 'info newts' for more information.\n"));

                    list_destroy (&nflist);
                    teardown ();

                    exit (EXIT_FAILURE);
                  }
                else if (parsed_time.tv_sec <= orig_seqtime)
                  {
                    orig_seqtime = seqtime = parsed_time.tv_sec;
                    if (sequencer == NONE)
                      sequencer = SEQUENCER;
                    alt_time = TRUE;
                  }
                else
                  {
                    fprintf (stderr, _("%s: parsed time '%s' in the future\n"),
                             program_name, optarg);
                    fprintf (stderr, _("See 'info newts' for more information.\n"));

                    list_destroy (&nflist);
                    teardown ();

                    exit (EXIT_FAILURE);
                  }
              }
            else
              {
                fprintf (stderr, _("%s: error parsing time '%s'\n"),
                         program_name, optarg);

                list_destroy (&nflist);
                teardown ();

                exit (EXIT_FAILURE);
              }

            break;
          }

        case 's':
          sequencer = SEQUENCER;
          break;

        case 't':
        case 'y':   /* Pseudo-option for --imsa */
          traditional = TRUE;
          if (!override_flag)
            seq_own_notes = TRUE;
          break;

        case 'u':
          {
            struct passwd *pw;

            if (getuid () != 0)
              {
                fprintf (stderr, _("%s: only root is allowed to use '--user'\n"),
                         program_name);

                list_destroy (&nflist);
                teardown ();

                exit (EXIT_FAILURE);
              }

            pw = getpwnam (optarg);
            if (pw)
              {
                seteuid (pw->pw_uid);

                newts_free (username);
                username = newts_strdup (pw->pw_name);

                if (alt_sequencer)
                  {
                    newts_nrealloc (seqname,
                                    strlen (username) + strlen (alt_seqname) + 2,
                                    sizeof (char));
                    strcpy (seqname, username);
                    strcat (seqname, ":");
                    strcat (seqname, alt_seqname);
                  }
                else
                  {
                    newts_free (seqname);
                    seqname = newts_strdup (username);
                  }
              }

            else
              {
                fprintf (stderr, _("%s: no such user: '%s'\n"), program_name,
                         optarg);

                list_destroy (&nflist);
                teardown ();

                exit (EXIT_FAILURE);
              }
          }
          break;

        case 'w':
          white_basenotes = TRUE;
          break;

        case 'x':
          sequencer = EXTENDED;
          break;

        case 'z':
          no_blacklist = TRUE;
          break;

        case 'h':
          printf (_("Usage: %s [OPTION]... NOTESFILE...\n"
                    "Run the UIUC-compatible notesfile client.\n\n"),
                  program_name);

          printf (_("If an argument to a long option is mandatory, it is also mandatory "
                    "for the\ncorresponding short option.\n\n"));

          printf (_("General options:\n"
                    "  -f, --file=FILE         Read list of notesfiles to view from specified file\n"
                    "  -g, --no-signature      Turn off automatic signature inclusion\n"
                    "  -h, --help              Display this help and exit\n"
                    "  -u, --user=USER         As root, run notes as the specified user\n"
                    "      --debug             Print debugging messages to stderr\n"
                    "      --version           Display version information and exit\n\n"));

          printf (_("Display options:\n"
                    "  -m, --modern            Use modern, consistent display style (default)\n"
                    "  -t, --traditional       Use traditional UIUC-style display\n"
                    "      --imsa              Use IMSA-style display (same as -t)\n\n"));

          printf (_("Blacklist options:\n"
                    "  -b, --black-threads     Skip threads with blacklisted basenotes while seqing\n"
                    "  -w, --white-basenotes   Do not apply blacklist to basenotes\n"
                    "  -z, --no-blacklist      Do not use the blacklist\n\n"));

          printf (_("Sequencer options:\n"
                    "  -a, --alternate=SEQ     Use alternate sequencer SEQ\n"
                    "  -e, --seq-own-notes     Make the sequencer view your notes\n"
                    "  -i, --index             Use the index sequencer\n"
                    "  -k, --skip-own-notes    Make the sequencer ignore your notes (default)\n"
                    "  -n, --no-sequencer      Do not use the sequencer (default)\n"
                    "  -o, --time=TIME         Use this date and time for the sequencer\n"
                    "  -s, --sequencer         Use the sequencer to read notes\n"
                    "  -x, --extended          Use the extended sequencer\n\n"));

          printf (_("Report bugs to <%s>.\n"), PACKAGE_BUGREPORT);

          list_destroy (&nflist);
          teardown ();

          if (fclose (stdout) == EOF)
            error (EXIT_FAILURE, errno, _("error writing output"));

          exit (EXIT_SUCCESS);

        case '?':
          fprintf (stderr, _("Try '%s --help' for more information.\n"),
                   program_name);

          list_destroy (&nflist);
          teardown ();

          exit (EXIT_FAILURE);
        }
    }

  if (optind == argc && !file_flag)
    {
      if (sequencer == NONE)
        {
          struct stat statbuf;

          fprintf (stderr, _("%s: too few arguments\n"), program_name);
          fprintf (stderr, _("Try '%s --help' for more information.\n"),
                   program_name);

          /* FIXME: should be a call to the notes system. */

          if (stat ("/etc/avail.notes", &statbuf) == 0)
            {
              fprintf (stderr,
                       _("Hit <RET> for a list of notesfiles on this system.\n"));
              getchar ();

              spawn_process (NULL, pager, "/etc/avail.notes", NULL);
            }

          list_destroy (&nflist);
          teardown ();

          exit (EXIT_FAILURE);
        }
      else
        {
          char *copy, *list, *token, *nfseq;

          nfseq = getenv ("NFSEQ");
          if (nfseq == NULL)
            {
              fprintf (stderr, _("%s: NFSEQ environment variable not set\n"),
                       program_name);
              fprintf (stderr, _("See 'info newts' for more information.\n"));

              list_destroy (&nflist);
              teardown ();

              exit (EXIT_FAILURE);
            }

          copy = newts_strdup (nfseq);

          token = strtok_r (copy, ", ", &list);
          while (token != NULL)
            {
              parse_nf (token, &nflist);
              token = strtok_r (NULL, ", ", &list);
            }

          newts_free (copy);
        }
    }
  else
    {
      while (optind < argc)
        parse_nf (argv[optind++], &nflist);
    }

  handle_signals ();

  /* For each notesfile, start up the master routine and go. */

  {
    ListNode *node = list_head (&nflist);

    while (node && !quit_flag)
      {
        newts_nfref *ref = (newts_nfref *) list_data (node);

        result = master (ref);

        node = list_next (node);

        if (result == QUITSEQ || result == QUITNOSEQ)
          quit_flag = TRUE;
      }
  }

  ignore_signals ();
  exit_curses ();

  if (*messages != '\0')
    printf ("%s", messages);

  teardown ();

  if (fclose (stdout) == EOF)
    error (EXIT_FAILURE, errno, _("error writing output"));

  exit (EXIT_SUCCESS);
}
